enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,477 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+Should we enable the [`ServiceAccountIssuerDiscovery` feature gate](https://github.com/kubernetes/enhancements/issues/1393) (disabled by default, alpha 1.18)?
+
+Pros:
+- Eliminates the need for external hosting of the OIDC discovery and keys documents
+- Eliminates the need for a controller to sync the keys when the bound ServiceAccount signer rotates
+
+Cons:
+- Requires enabling alpha feature
+- Does not work if kube-apiserver is not publicly accessible i.e. accessible
+  from AWS STS.  We would need to document the process for mirroring to an
+  external location like an S3 bucket in that case.
+
+## Summary
+
+When deployed on AWS and configured to use [bound service
+tokens](https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/bound-sa-tokens.md),
+Kubernetes pods can assume Roles managed by IAM via a binding created by a
+ServiceAccount annotation.  AWS has a
+[webhook](https://github.com/aws/amazon-eks-pod-identity-webhook/) they maintain
+for EKS that does enables this ability.  This enhancement integrates the same
+functionality into OpenShift.
+
+The `cloud-credential-operator` will be responsible for managing this new
+ability as it serves as an alternative credential _injection_ mechanism versus
+the existing mechanism of injecting an IAM user access ID and key into a
+secret.  The existing role/user _creation_ via a CredentialRequest CR is common
+between the methods.
+
+The new injection mechanism works by creating a `ServiceAccount` annotated with
+a IAM role the pod should assume. When a pod runs with the annotated
+`ServiceAccount`, the webhook injects a projected volume into the pod that
+contains the bound token that can be used to assume the IAM role.  The AWS SDK
+searches for this token automatically.
+
+## Motivation
+
+This allows users to more cleanly express and manage pod identities in IAM when
+running on AWS.  The use of bound tokens to assume roles (vs user id/key) is
+preferred as these tokens can expire and be rotated.
+
+### Goals
+
+1. A pod can assume an IAM role created with a `CredentialRequest` CR using a
+   bounded token
+2. An admin can create a IAM role via `CredentialRequest` CR and inject to a
+   `ServiceAccount` (just as they can create an IAM user and inject to a
+   `Secret` currently)
+4. The token signing CA (`bound-sa-token-signing-certs`) can rotate and should
+   sync between OCP and the OIDC provider in IAM.  Rotation should be seamless.
+   Assumes applications in pods monitor the token for rotation (AWS SDK does
+   this)
+
+### Non-Goals
+
+1. Convert existing IAM user-based components (e.g. ingress, image-registry,
```

Comment:
Do we plan to update the existing CredRequests to use this new mechanism on AWS? And if so, how can we upgrade without triggering something bad here. These CredRequests are delivered via the release manifest and I believe that means the CVO will update them to their new state. Is there a way we can make this manual and opt-in? 

In doing so, the operator permissions would then become fixed and need manual intervention in the event an upgraded operator needed new permissions.

Reply:
No we are not planning to update existing `CredentialRequests` at this time.

For this stage, I recommended we disallow CR changes where the spec changes from a `secretRef` to a `serviceAccountRef` or visa versa.

There doesn't seem to be a way to mark CRD fields as immutable atm https://github.com/kubernetes/kubernetes/issues/65973

We'll probably just overwrite the change in the controller is someone tries to do that.

Reply:
> We'll probably just overwrite the change in the controller is someone tries to do that.

Do we have an admission webhook for this?  What happens if someone tries to push a change while the cred minter is, for some reason, out of action?  The new cred minter comes up and... searches for existing Secrets to decide if it needs to overwrite a `serviceAccountRef`?